torchmultimodal/modules/layers/transformer.py

```diff
@@ -328,11 +334,68 @@ def forward(
             )
 
 
-def _apply_layernorm(x: Tensor, layernorm: nn.Module) -> Tensor:
-    """Supports mixed-precision training by casting to fp32 for layernorm and back"""
-    if x.dtype != torch.float32:
-        x_fp32 = x.float()
-        x_fp32 = layernorm(x_fp32)
-        return x_fp32.type_as(x)
-    else:
-        return layernorm(x)
+class TransformerEncoder(nn.Module):
+    def __init__(
+        self,
+        n_layer: int,
+        d_model: int,
+        n_head: int,
+        dim_feedforward: int,
+        dropout: float = 0.0,
+        activation: Callable[..., nn.Module] = nn.ReLU,
+        layer_norm_eps: float = 1e-12,
+        norm_first: bool = False,
+    ):
+        super().__init__()
+        self.layer = nn.ModuleList(
+            [
+                TransformerEncoderLayer(
+                    d_model,
+                    n_head,
+                    dim_feedforward,
+                    dropout,
+                    activation,
+                    layer_norm_eps,
+                    norm_first,
+                )
+                for _ in range(n_layer)
+            ]
+        )
+        self.num_layers = n_layer
+
+    def forward(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerOutput:
+        all_hidden_states: Tuple[Tensor, ...] = () if return_hidden_states else None
+        all_self_attentions: Tuple[Tensor, ...] = () if return_attn_weights else None
+
+        for i, layer_module in enumerate(self.layer):
+            if return_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
```

Comment:
hm will this add the original input hidden states too?

Reply:
this is straight from the original `FLAVATransformerEncoder`: https://github.com/facebookresearch/multimodal/blob/f9327dec04a5e8c824fba36e4ed52528e44541f6/torchmultimodal/modules/layers/transformer.py#L384